### PR TITLE
sql: fix EXPLAIN (DISTSQL) for CREATE STATISTICS

### LINKED ISSUE #36677: stats: fix EXPLAIN (DISTSQL) for CREATE STATISTICS
`EXPLAIN (DISTSQL)` used to show the DistSQL plan for `CREATE STATISTICS`, which could be useful for debugging purposes. Ever since `CREATE STATISTICS` was converted to use the jobs framework, `EXPLAIN (DISTSQL)` only shows the local node corresponding to the function that starts the job.  We should fix this so `EXPLAIN (DISTSQL)` shows the actual DistSQL plan created for the job.

### PR DESCRIPTION
**Pull request by rytaft**
`EXPLAIN (DISTSQL)` has been broken for `CREATE STATISTICS` ever since
`CREATE STATISTICS` was converted to use the jobs framework. Instead
of showing the DistSQL plan, it has only shown the local node
corresponding to the function that started the job. This commit fixes
`EXPLAIN (DISTSQL)` for `CREATE STATISTICS` so it shows the actual
DistSQL plan used for the job.

Fixes #36677

Release note (sql change): EXPLAIN (DISTSQL) CREATE STATISTICS now
shows the DistSQL plan used by the CREATE STATISTICS job.

### CODE REVIEW

**Comment by jordanlewis** [comment]
Thanks for doing this! At a high level, I do wonder whether this is worth the added special case complexity. Do we expect this case to actually come up a lot, that people want to see the `EXPLAIN(DISTSQL)` for a create statistics job?

Also, what happens with `EXPLAIN (ANALYZE, DISTSQL)`? Probably that won't work here, which is fine, but it's worth thinking about.

If the use case here is really internal, might I suggest adding a vmodule flag that logs the distsql plan from the `CREATE STATISTICS` job instead?

**Comment by jordanlewis** [comment]
Another thing to point out is that the other jobs that make distsql plans don't do the "right" thing here either, and we've been okay with that. IMPORT, for example.

A more holistic fix (not that I'm suggesting you do that this second) could be to add infrastructure for all distributed-job-creating planNodes to get a way to expose a better looking EXPLAIN(DISTSQL).

**Comment by jordanlewis** [comment]
I meant `EXPLAIN ANALYZE (DISTSQL)`, not whatever I said above.

**Review by rytaft — COMMENTED** [review]
Thanks for the comments, @jordanlewis! 

> Do we expect this case to actually come up a lot, that people want to see the EXPLAIN(DISTSQL) for a create statistics job?

I'm not sure, but I think it would be nice for users to have the option. For example, if a user notices that a stats run is impacting performance, this could be a tool to help understand why.

> Also, what happens with EXPLAIN ANALYZE (DISTSQL)?

It seems to be working -- is there a reason you thought it wouldn't work? I added a test case as well.

> A more holistic fix... could be to add infrastructure for all distributed-job-creating planNodes

That's a good idea. I changed the code to use an interface instead of making a special case for CREATE STATS. Let me know what you think.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36687#-:-Lc6TG9T4xzgBMCrYvfR:b-sw82rx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, and @RaduBerinde)

File: pkg/sql/explain_distsql.go [hunk 1]
```diff
@@ -142,10 +158,17 @@ func (n *explainDistSQLNode) startExec(params runParams) error {
 		}
 	}
 
-	plan, err := distSQLPlanner.createPlanForNode(planCtx, n.plan)
+	var plan PhysicalPlan
+	var err error
+	if planNode, ok := n.plan.(distSQLExplainable); ok {
+		plan, err = planNode.makePlanForExplainDistSQL(planCtx, distSQLPlanner)
```
Comment (jordanlewis) [inline pkg/sql/explain_distsql.go]: This won't work in the case of something like `SELECT blah FROM [CREATE STATISTICS...]`, assuming `CREATE STATISTICS` returns a table, which is probably fine but it's worth pointing out, I think.

**Review by jordanlewis — APPROVED** [review]
Changes to EXPLAIN and so on LGTM - probably should get a review on the statistics change as well. Thanks for making the change more general!

**Review by rytaft — COMMENTED** [review]
TFTR! I'll wait for @RaduBerinde or @justinj to review the changes to `create_stats.go`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36687#-:-Lc6lQE7Erf7Dvvst7T1:b-ga36kv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @justinj, and @RaduBerinde)

___
*[pkg/sql/explain_distsql.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36687#-Lc37gltdf8UBv160qdy-r1-164:-Lc6knkNBzzdKl6oLK5j:b-oquev7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0d31e7c17661e085234b9ef68c7da0f97f0bb8d/pkg/sql/explain_distsql.go#L164)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This won't work in the case of something like `SELECT blah FROM [CREATE STATISTICS...]`, assuming `CREATE STATISTICS` returns a table, which is probably fine but it's worth pointing out, I think.
</blockquote></details>

Yea that's ok -- `CREATE STATISTICS` is a DDL statement so it returns nothing -- so people shouldn't be running that anyway (they can use `SHOW STATISTICS` to actually view the stats)

**Review by RaduBerinde — COMMENTED** [review]
LGTM. Nice change!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36687#-:-Lc78P9yCZS7lPZ5jeil:b-l9d7k7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, @justinj, and @RaduBerinde)

**Review by rytaft — COMMENTED** [review]
Thanks!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36687#-:-Lc7F8QkBBIV9EPSeG0I:b-voiyyw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jordanlewis, and @justinj)

**Pull request merged**